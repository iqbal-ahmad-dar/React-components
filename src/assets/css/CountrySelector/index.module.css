.label {
  @apply relative !mb-0 inline-flex items-center py-[5px] text-sm font-semibold text-dark;
}
.formControl {
  @apply flex items-center h-[32px] w-full rounded-sm border border-solid border-border bg-white px-2.5 py-0 text-sm/[1.5] text-[14px] font-normal text-dark shadow-none placeholder:text-secondary focus:border-focus focus:outline-none;
}
.dropdownCard {
  @apply absolute top-[31px] z-50 max-h-[200px] w-full overflow-x-hidden overflow-y-scroll border border-solid border-border bg-white;
}
.focusedBorder {
  @apply !border !border-solid !border-black;
}
.flagContainer {
  @apply flex !h-6 w-6 items-center;
}
.flagImage {
  @apply w-[20.5px] h-5;
}
.expandIcon {
  @apply !-mr-1;
}
.dropdownItem {
  @apply flex h-8 w-full cursor-pointer items-center px-[10px] py-[5px] hover:bg-[#0000000d];
}
.dropdownFlagContainer {
  @apply flex h-5 w-6 items-center;
}
.dropdownFlagImage {
  @apply w-[20.5px] !h-5;
}
.dropdownCountryName {
  @apply text-sm font-normal text-dark;
}
