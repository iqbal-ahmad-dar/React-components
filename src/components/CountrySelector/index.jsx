import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import { countriesList } from './countriesList';
import style from '../../assets/css/CountrySelector/index.module.css';
const propTypes = {
	label: PropTypes.string,
	onChange: PropTypes.func,
	defaultCountry: PropTypes.object,
};
const CountrySelector = ({ label, onChange, defaultCountry }) => {
	const [countries, setCountries] = useState(countriesList);
	const defaultCountryIndex = countries.findIndex(
		(country) =>
			country?.alpha2?.toLowerCase() ===
			defaultCountry?.alpha2?.toLowerCase()
	);
	const [selectedCountry, setSelectedCountry] = useState(
		defaultCountryIndex !== -1
			? countries[defaultCountryIndex]
			: countries[103]
	);
	const [filterText, setFilterText] = useState('');
	const [focus, setFocus] = useState(false);
	const [showDropdown, setShowDropdown] = useState(false);
	const divRef = useRef(null);
	const closeRef = useRef(null);
	const handleContentChange = (event) => {
		setFilterText(event.target.innerText);
	};
	const toggleDropdown = () => {
		setShowDropdown((prev) => !prev);
	};

	const handleCountryClick = (country) => {
		let newCountry = {
			country: country.name,
			flag: country.alpha2,
			alpha2: country.alpha2,
		};
		setSelectedCountry(country);
		onChange(newCountry);
		setFocus(true);
		setFilterText('');
		setShowDropdown(false);
	};

	useEffect(() => {
		onChange(selectedCountry);
		const handleClickOutside = (event) => {
			if (closeRef.current && !closeRef.current.contains(event.target)) {
				setShowDropdown(false);
				setFocus(false);
			}
		};
		document.addEventListener('mousedown', handleClickOutside);
		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, []);

	return (
		<div className="relative mb-2.5">
			{label && <div className={style.label}>{label}</div>}
			<div ref={closeRef} style={{ position: 'relative' }}>
				<div
					className={`${style.formControl} ${focus ? style.focusedBorder : ''}`}
					onClick={toggleDropdown}
					tabIndex={0}
				>
					<div className="relative flex w-full items-center gap-1.5 h-full">
						<div className={style.flagContainer}>
							<img
								src={require(`./images/${selectedCountry.alpha2}.png`)}
								alt="img"
								className={style.flagImage}
							/>
						</div>
						<div className="flex w-full items-center focus:border-none focus:outline-none">
							<div
								className="w-full focus:border-none focus:outline-none"
								contentEditable={true}
								ref={divRef}
								onInput={handleContentChange}
								suppressContentEditableWarning={true}
							>
								{selectedCountry.name}
							</div>
							<ExpandMoreIcon className={style.expandIcon} />
						</div>
					</div>
				</div>
				{showDropdown && (
					<div className={style.dropdownCard}>
						{countries
							.filter((country) =>
								country.name
									.toLowerCase()
									.includes(filterText.toLowerCase())
							)
							.map((country, index) => (
								<div
									key={index}
									onClick={() => handleCountryClick(country)}
									className={style.dropdownItem}
								>
									<div className="flex items-center gap-1.5">
										<div className={style.dropdownFlagContainer}>
											<img
												src={require(`./images/${country.alpha2}.png`)}
												alt={country.country}
												className={style.dropdownFlagImage}
											/>
										</div>
										<span className={style.dropdownCountryName}>
											{country.name}
										</span>
									</div>
								</div>
							))}
					</div>
				)}
			</div>
		</div>
	);
};

CountrySelector.propTypes = propTypes;

export default CountrySelector;
