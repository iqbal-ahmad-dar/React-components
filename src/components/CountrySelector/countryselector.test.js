import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import CountrySelector from "../CountrySelector";
import { countriesList } from "../CountrySelector/countriesList";
describe("CountrySelector Component", () => {
  const onChangeMock = jest.fn();
  const defaultCountry = countriesList[103];
  const renderComponent = (props = {}) => {
    return render(
      <CountrySelector
        label="Select Country"
        onChange={onChangeMock}
        defaultCountry={defaultCountry}
        {...props}
      />
    );
  };
  beforeEach(() => {
    onChangeMock.mockClear();
  });
  test("renders correctly", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });
  test("displays the label when provided", () => {
    renderComponent();
    expect(screen.getByText("Select Country")).toBeInTheDocument();
  });
  test("displays the default country", () => {
    renderComponent();
    expect(screen.getByText(defaultCountry.name)).toBeInTheDocument();
  });
  test("opens the dropdown on click", () => {
    renderComponent();
    const formControl = screen.getByText(defaultCountry.name).parentElement
      .parentElement;
    fireEvent.click(formControl);
    expect(screen.getByText(countriesList[0].name)).toBeInTheDocument();
  });
  test("closes the dropdown when clicking outside", () => {
    renderComponent();
    const formControl = screen.getByText(defaultCountry.name).parentElement
      .parentElement;
    fireEvent.click(formControl);
    fireEvent.mouseDown(document);
    expect(screen.queryByText(countriesList[0].name)).not.toBeInTheDocument();
  });
});
