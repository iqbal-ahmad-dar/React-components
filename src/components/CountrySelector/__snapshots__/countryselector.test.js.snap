// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountrySelector Component renders correctly 1`] = `
<div>
  <div
    class="relative mb-2.5"
  >
    <div
      class="label"
    >
      Select Country
    </div>
    <div
      style="position: relative;"
    >
      <div
        class="formControl "
        tabindex="0"
      >
        <div
          class="relative flex w-full items-center gap-1.5 h-full"
        >
          <div
            class="flagContainer"
          >
            <img
              alt="img"
              class="flagImage"
              src="in.png"
            />
          </div>
          <div
            class="flex w-full items-center focus:border-none focus:outline-none"
          >
            <div
              class="w-full focus:border-none focus:outline-none"
              contenteditable="true"
            >
              India
            </div>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium expandIcon css-i4bv87-MuiSvgIcon-root"
              data-testid="ExpandMoreIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M16.59 8.59 12 13.17 7.41 8.59 6 10l6 6 6-6z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
